torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -665,16 +669,20 @@ void Reducer::finalize_bucket_dense(Bucket& bucket) {
         local_used_maps_reduced_ = true;
       }
 
-      auto bucket_view =
-          replica.contents.narrow(0, offset, length).view(variable.sizes());
+      auto bucket_slice =
+          replica.contents.narrow(0, offset, length);
       auto& grad = variable.grad();
 
       // If a parameter is globally unused, we keep its grad untouched.
       if (!global_unused) {
         if (!grad.defined()) {
-          grad = at::empty(bucket_view.sizes(), bucket_view.options());
+          grad = at::empty(variable.sizes(), bucket_slice.options());
```

Comment:
I'm not sure in what situation this will be triggered, but it will create nchw contiguous grad. Is it possible that grad is defined in make_variable_copy_dense, grad is permuted there, but grad is not defined here? If this scenario is possible it will lead to errors.

Reply:
That's a good point and was also a concern of mine.  However, DDP hooks are post-accumulate hooks, which means if the grad was going to be created by this backward pass, it has been by this point.  In other words, the only way a grad is `!defined` is if it did not receive a grad this iteration.  In this case, to avoid the good old nccl hang issue if other processes DID created a grad for it, Reducer fills that undefined grad's bucket_slice with zeros when populating the buckets before kicking off the allreduce ([philosophy](https://github.com/pytorch/pytorch/blob/6dfb14fbefb8d3045f99c84d5bc579b0e41ca37e/torch/csrc/distributed/c10d/reducer.cpp#L219-L225), [code](https://github.com/pytorch/pytorch/blob/6dfb14fbefb8d3045f99c84d5bc579b0e41ca37e/torch/csrc/distributed/c10d/reducer.cpp#L246)).

So yeah, we might end up copying out an incorrect permutation if we hit this code path, but it will only copy zeros.

That being said, we still might end up creating a grad whose sizes match the variable but whose contiguity pattern doesn't.  I think what we really want here is for grad to be created with sizes _and strides_ matching `variable`.  `empty_like` might do this, I'll double check.  What do you think?

Reply:
Update: `empty_like` is stride preserving.

Reply:
What you said above deserves to be put in a comment. "Yes, we are filling it with incorrect permutation but no one cares because its zeros anyway" is such a hollow and fragile explanation that it has to be in the codebase next time someone has to debug ddp issues.

Reply:
Agreed, especially since it creates an implicit contract between the point in the code where the allreduces are kicked off and the point in the code where allreduces are unbucketed.

Reply:
I think if we create the grad using empty_like we won't be filling via the wrong permutation, but should still comment about relationship between bucketing and unbucketing.